### PR TITLE
Integration test cargo feature that spawns `bitcoind` within rust

### PR DESCRIPTION
**Pull request by evanlinjin**
### Description

We introduce the `bitcoind` cargo feature for integration tests.

This cargo feature adds the ability for integration tests to spawn a child process of `bitcoind` (via the `bitcoind` crate).

When the `bitcoind` cargo feature is enabled, internally-spawned `bitcoind` will replace mock rpc client and rpc client that interacts with externally-spawned `bitcoind`.

A new env `BITCOIN_EXE` is added which the `bitcoind` binary location can be specified. If unspecified, the `bitcoind` crate will download the binary.

`RealBitcoinTest` is now an enum, either specifying just a client, or both a service and a client (with `bitcoind::BitcoinD`).

### Checklist
- [x] Update README

### CODE REVIEW

File: integrationtests/tests/fixtures/mod.rs [hunk 1]
```diff
@@ -116,60 +117,161 @@ pub async fn fixtures(
...
-                lightning.gateway_node_pub_key,
-            )
-            .await;
+    let env_disable_mocks = env::var("FM_TEST_DISABLE_MOCKS").unwrap_or_default();
+    match env_disable_mocks.as_str() {
+        "1" => {
+            #[cfg(not(feature = "bitcoind"))]
+            info!("Testing with EXTERNAL bitcoind and EXTERNAL cln");
+            #[cfg(feature = "bitcoind")]
+            info!("Testing with INTERNAL bitcoind and EXTERNAL cln");
```
Comment (elsirion) [inline integrationtests/tests/fixtures/mod.rs:126]: I don't think we want the combinatorial explosion here since we'd need to test all combinations.

**Review by elsirion — CHANGES_REQUESTED** [review]
I think this needs re-evaluation on a high level. I don't see the difference in outcome between spawning bitcoind in rust and running it internally. What I do agree with is that the tests should be more self-contained and probably not rely on externally started services (if that doesn't have other downsides, @jkitman might know).

So instead of adding this as another way of running bitcoind this should become the only way of running bitcoind in the rust integration tests (the shell tests will be unaffected). To minimize inconsistency core-lightning should be started in a similar fashion.

**Comment by evanlinjin** [comment]
@elsirion Thank you for your review! Just to make sure I understand you, this should be my next steps with this PR:
* Make running `bitcoind` via rust the only way of running `bitcoind` in rust integration tests.
* Run core lightning in the same fashion.

Is this correct?

**Comment by elsirion** [comment]
Yes, that's what I meant. I'd like to hear @jkitman's opinion since he built the rust integration tests, but I think starting the daemons externally is a historic artifact that we haven't gotten around to fix.

File: integrationtests/README.md [hunk 2]
```diff
@@ -71,3 +71,15 @@ If you wish to clean-up the services you either exit the nix-shell or run:
...
+
+## Spawn `bitcoind` service from Rust
+When the `bitcoind` cargo feature is set, `bitcoind` is run by the Rust integration tests instead of the mock implementation (with `FM_TEST_DISABLE_MOCKS` not set as `1`) or an external `bitcoind` service (with `FM_TEST_DISABLE_MOCKS=1`).
+
+The `BITCOIN_EXE` can be defined to the local location of a `bitcoind` binary. If `BITCOIN_EXE` is not set, the binary will be downloaded.
+
+```shell
+# Run integration tests with internally-ran bitcoind and mock core lightning node
+export FM_TEST_DISABLE_MOCKS=0
+export BITCOIN_EXE=$(which bitcoind) # We use our own bitcoind binary
```
Comment (dpc) [inline integrationtests/README.md:83]: `s/EXE/BIN/`?

File: integrationtests/tests/fixtures/fake.rs [hunk 3]
```diff
@@ -86,6 +86,7 @@ pub struct FakeBitcoinTest {
 }
 
 impl FakeBitcoinTest {
+    #[allow(dead_code)]
```
Comment (jkitman) [inline integrationtests/tests/fixtures/fake.rs:89]: Why is this dead code now?

**Comment by jkitman** [comment]
@evanlinjin So the advantage of using `bitcoind` is that it can be automatically downloaded and run from Rust, rather than to rely on it already running?

I would also agree with @elsirion we shouldn't have the feature flags and just switch the real fixtures to use `bitcoind` and the fake fixtures continue to use the mock bitcoin impl.

So long as all the tests can pass with `bitcoind` I don't see a problem to switching to use it.

**Comment by evanlinjin** [comment]
> @evanlinjin So the advantage of using `bitcoind` is that it can be automatically downloaded and run from Rust, rather than to rely on it already running?

Yes, this is correct.

> I would also agree with @elsirion we shouldn't have the feature flags and just switch the real fixtures to use `bitcoind` and the fake fixtures continue to use the mock bitcoin impl.

Would it make sense to NOT use mock `bitcoind` service for testing? It means less code we need to maintain, and we can just test with the real thing.

Also the reason why I switched over to using feature flags is because nix did not seem to like the [`bitcoind`](https://docs.rs/bitcoind/latest/bitcoind/) crate - making it a dev dependency caused nix to claim the hashes does not add up. I do not understand nix so if I can get some guidance in this regard, it will be appreciated.

> We may also need to change the `scripts/rust-tests.sh` to avoid starting two instances of `bitcoind`.

That sounds good, I will see how I can get that done!

**Comment by jkitman** [comment]
> Would it make sense to NOT use mock bitcoind service for testing

@evanlinjin the purpose of the mock was for faster testing and eliminating any shared state between tests so they can be run concurrently.  If we use `bitcoind` it will likely make tests fail if run concurrently.

**Comment by dpc** [comment]
I'm marking this as blocked, as it's unclear if we can proceed with this one. Feel free to keep discussing etc. but it just makes easier for us to know which things we need to actively look into and which not.

> Also the reason why I switched over to using feature flags is because nix did not seem to like the [bitcoind](https://docs.rs/bitcoind/latest/bitcoind/) crate - making it a dev dependency caused nix to claim the hashes does not add up. I do not understand nix so if I can get some guidance in this regard, it will be appreciated.

I would be happy to help, but need more details. If you want you can ping me on Discord, or just paste some more info here.

**Comment by justinmoon** [comment]
Closing this because we’ve recently started spawning everything via rust

**Pull request closed without merging**